Mathlib/RingTheory/TwoSidedIdeal.lean

```diff
@@ -0,0 +1,294 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import Mathlib.RingTheory.Congruence
+import Mathlib.RingTheory.Ideal.Basic
+import Mathlib.Algebra.Module.LinearMap.Basic
+import Mathlib.Algebra.Module.Opposites
+
+/-!
+# Two Sided Ideals
+
+In this file, for any `Ring R`, we reinterpret `I : RingCon R` as a two-sided-ideal of a ring.
+
+## Main definitions and results
+* `RingCon.setLike`: Every `I : RingCon R` can be interpreted as a set of `R` where `x ∈ I` if and
+  only if `I x 0`. In this sense, two-sided-ideals of `R` are exactly `RingCon R`.
+* `RingCon.orderIsoRingConMop`: The two-sided-ideals of `R` and that of `Rᵐᵒᵖ` correspond
+  bijectively to each other.
+
+* `RingCon.addCommGroup`: when viewing `I : RingCon R` as a set, it is an abelian group.
+* `RingCon.leftModule`: when viewing `I : RingCon R` as a set, it is a left module over `R`.
+* `RingCon.rightModule`: when viewing `I : RingCon R` as a set, it is a right module over `R`.
+
+* `RingCon.span`: let `s ⊆ R` be a subset, `span s` is the smallest two-sided-ideal containing `s`.
+
+## Notes
+`SetLike (RingCon R) R` makes sense for any `NonUnitalNonAssocRing R`.
+But later for the `module` part, we will assume `R` is a ring.
+
+-/
+
+open MulOpposite
+
+section ideal
+
+namespace RingCon
+
+section NonUnitalNonAssocRing
+
+variable {R : Type*} [NonUnitalNonAssocRing R] (I : RingCon R)
+
+instance setLike : SetLike (RingCon R) R where
+  coe t := {r | t r 0}
+  coe_injective' t₁ t₂ (h : {x | _} = {x | _}) := by
+    refine RingCon.ext fun a b ↦ ⟨fun H ↦ ?_, fun H ↦ ?_⟩
+    · have H' : a - b ∈ {x | t₁ x 0} := sub_self b ▸ t₁.sub H (t₁.refl b)
+      rw [h] at H'
+      convert t₂.add H' (t₂.refl b) using 1 <;> abel
+    · have H' : a - b ∈ {x | t₂ x 0} := sub_self b ▸ t₂.sub H (t₂.refl b)
+      rw [← h] at H'
+      convert t₁.add H' (t₁.refl b) using 1 <;> abel
+
+lemma mem_iff (x : R) : x ∈ I ↔ I x 0 := Iff.rfl
+
+lemma rel_iff (x y : R) : I x y ↔ x - y ∈ I := by
+  rw [mem_iff]
+  constructor
+  · intro h; convert I.sub h (I.refl y); abel
+  · intro h; convert I.add h (I.refl y) <;> abel
+
+/--
+Any two-sided-ideal in `A` corresponds to a two-sided-ideal in `Aᵒᵖ`.
+-/
+@[simps]
+def toMop (rel : RingCon R) : (RingCon Rᵐᵒᵖ) :=
```

Comment:
Following other structures, I'd call this `RingCon.op` and the next construction `RingCon.unop`. They probably below to `Mathlib.RingTheory.Congruence` too.

Reply:
done in [d107a35](https://github.com/leanprover-community/mathlib4/pull/13401/commits/d107a356f2cc703f2eb856fe937bb99d4540285f)